predictor/tests/test_register.py

```diff
@@ -0,0 +1,247 @@
+
+    def test_incorrect_email(self):
+        context = correct_user.copy()
+        context['email'] = 'email'
+
+        resp = self.client.post(reverse('predictor:register'), context)
+
+        self.assertEqual(resp.status_code, 200)
+        self.assertTemplateUsed(resp, 'predictor/register.html')
+
+        self.assertTrue('incorrect_email' in resp.context)
+        self.assertTrue(resp.context['incorrect_email'])
+
+        for field in saved_fields:
+            self.assertTrue(field in resp.context)
+            self.assertEqual(resp.context[field], context[field])
+
+    def test_correct_registration(self):
+        context = correct_user.copy()
+
+        resp = self.client.post(reverse('predictor:register'), context,
+                                follow=True)
+
+        self.assertRedirects(resp, reverse('predictor:index'))
+
+        self.assertEqual(len(User.objects.filter(username='test-login1')), 1)
+
+        user = User.objects.filter(username='test-login1')[0]
+
+        self.assertEqual(user.username, 'test-login1')
+        self.assertEqual(user.first_name, 'oleg')
+        self.assertEqual(user.last_name, 'tester')
+        self.assertEqual(user.email, 'oleg@yandex.ru')
+        self.assertTrue(user.check_password('OlEgCOdEr'))
+
+        algo_settings = AlgorithmSettings.objects.filter(user=user)[0]
+
+        self.assertEqual(algo_settings.question, correct_user['question'])
+        self.assertEqual(algo_settings.answer, correct_user['answer'])
+
+    def test_used_email(self):
+        context = correct_user.copy()
+        context['email'] = 'lion@mail.ru'
+
+        resp = self.client.post(reverse('predictor:register'), context)
+
+        self.assertEqual(resp.status_code, 200)
+        self.assertTemplateUsed(resp, 'predictor/register.html')
+
+        self.assertTrue('incorrect_email' in resp.context)
+        self.assertTrue(resp.context['incorrect_email'])
+
+        for field in saved_fields:
+            self.assertTrue(field in resp.context)
+            self.assertEqual(resp.context[field], context[field])
+
+    def test_used_login(self):
+        context = correct_user.copy()
+        context['login'] = 'test-login0'
+
+        resp = self.client.post(reverse('predictor:register'), context)
+
+        self.assertEqual(resp.status_code, 200)
+        self.assertTemplateUsed(resp, 'predictor/register.html')
+
+        self.assertTrue('incorrect_login' in resp.context)
+        self.assertTrue(resp.context['incorrect_login'])
+
+        for field in saved_fields:
+            self.assertTrue(field in resp.context)
+            self.assertEqual(resp.context[field], context[field])
+
+    def test_redirection_signed_in(self):
+        user = User.objects.get(username='test-login0')
+        self.client.force_login(user=user)
+
+        resp = self.client.get(reverse('predictor:auth'))
+
+        self.assertRedirects(resp, reverse('predictor:index'))
+
+
```

Comment:
There are lots of not redundant blank lines. Need only one in the end of the file.